### PR TITLE
feat: move cassettes to git submodule

### PR DESCRIPTION
**Pull request by BaseInfinity**
### Background
We want to be able to access all our cassettes from one directory/repo/submodule and keep the cassettes in sync with correct version of `master` on main repo.

Using git submodules allows us to achieve that without knowing exactly what commit/hash we need to be at. Git submodule will always update to latest version of `main` relative to the primary repo. 

This means that as long as we update our git submodules on merges into `master`, we will always be able to keep our cassettes in sync and even if we need to run older tests on older commits.

### Changes
- [x] Move all the cassettes to one directory
- [x] Move to git submodule
- [x] Update CI to work with cassettes git submodule 
- [x] Make sure we are handling updating of git submodules correctly when we switch branches (tap into git hooks)
- [x] Ensure on merge to master that we update `tests/cassettes` git submodule (delete old update code)
- [ ] Ensure somewhere in the install process that we create the symbolic link so devs can take advantage of `post-checkout` commit hook.  `ln -s hooks/post-checkout .git/hooks/post-checkout`

Closes https://github.com/Significant-Gravitas/Auto-GPT/issues/4284

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes. 
- [x] I have run `black .` and `isort .` against my code to ensure it passes our linter.

### CODE REVIEW

**Comment by BaseInfinity** [comment]
Hmm I'm not sure why it's complaining about conflicting files when I deleted that directory `tests/integration/challenges/memory/cassettes`

**Comment by Pwuts** [comment]
> Hmm I'm not sure why it's complaining about conflicting files when I deleted that directory `tests/integration/challenges/memory/cassettes`

You have (re)moved it in your fork, but it has been updated in `master`. You can resolve the conflict by updating the cassettes in their new location with the updated content that's in `master`.

**Comment by BaseInfinity** [comment]
> @Pwuts: You have (re)moved it in your fork, but it has been updated in `master`. You can resolve the conflict by updating the cassettes in their new location with the updated content that's in `master`.

derp, I'm not used to working outside the repo, old habits die hard lol. 

Thanks for the help. I synced my fork with `master` and merged it with my changes and re-ran the tests to generate cassettes in their correct directory with latest code.

**Comment by BaseInfinity** [comment]
> @Pwuts: You have (re)moved it in your fork, but it has been updated in `master`. You can resolve the conflict by updating the cassettes in their new location with the updated content that's in `master`. 

derp, I'm not used to working outside the repo, old habits die hard lol. Thanks for the help. I synced my fork with `master` and merged it with my changes and re-ran the tests to generate cassettes in their correct directory with latest code.

**Comment by BaseInfinity** [comment]
Looks like I'm failing tests, I'll have to poke into it tomorrow

File: tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml [hunk 1]
```diff
@@ -7,11 +7,11 @@ interactions:
...
-      with no legal complications.\n\nGOALS:\n\n1. Use the browse_website command
-      to visit http://books.toscrape.com/catalogue/meditations_33/index.html and answer
-      the question ''What is the price of the book?''\n2. Write the price of the book
-      to a file named \"browse_website.txt\".\n3. Use the task_complete command to
-      complete the task.\n4. Do not use any other commands.\n\n\nConstraints:\n1.
+      with no legal complications.\nThe OS you are running on is: macOS-12.6\n\nGOALS:\n\n1.
+      Use the browse_website command to visit http://books.toscrape.com/catalogue/meditations_33/index.html
+      and answer the question ''What is the price of the book?''\n2. Write the price
+      of the book to a file named \"browse_website.txt\".\n3. Use the task_complete
+      command to complete the task.\n4. Do not use any other commands.\n\n\nConstraints:\n1.
```
Comment (Pwuts) [inline tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml]: Looks like you have regenerated these cassettes with non-standard settings, which means they won't be usable for CI.

File: tests/conftest.py [hunk 2]
```diff
@@ -8,6 +8,12 @@
 from autogpt.llm import ApiManager
 from autogpt.workspace import Workspace
 
+
+@pytest.fixture(scope="module")
+def vcr_cassette_dir(request):
+    return os.path.join("tests/fixtures/cassettes", request.module.__name__)
```
Comment (Pwuts) [inline tests/conftest.py]: Why `tests/fixtures/cassettes`? There is nothing else in `fixtures` so it might as well be `tests/cassettes` :)

**Review by Pwuts — COMMENTED** [review]
Looks like a good start, although watch out when regenerating the cassettes. If they are generated with non-standard settings (e.g. `EXECUTE_LOCAL_COMMANDS=True`), they won't be used in CI.

File: tests/conftest.py (hunk 2 above)
Comment (BaseInfinity) [reply tests/conftest.py]: Lol true, i'll make the change

File: tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml (hunk 1 above)
Comment (BaseInfinity) [reply tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml]: > Looks like you have regenerated these cassettes with non-standard settings, which means they won't be usable for CI.

I think for now I'll just move them over, I didn't realize there's a particular set of settings needed when generating these cassettes

**Comment by BaseInfinity** [comment]
> Looks like a good start, although watch out when regenerating the cassettes. If they are generated with non-standard settings (e.g. `EXECUTE_LOCAL_COMMANDS=True`), they won't be used in CI.

Ahh gotcha okay, I'm going to make the suggested changes and instead just move the latest cassettes over from master instead of trying to regenerate them locally

File: tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml (hunk 1 above)
Comment (BaseInfinity) [reply tests/fixtures/cassettes/tests.integration.challenges.basic_abilities.test_browse_website/test_browse_website.yaml]: https://github.com/Significant-Gravitas/Auto-GPT/pull/4294/commits/c917a4fbb87c9efa5a6800456dbc72e18fe503e8

File: tests/conftest.py (hunk 2 above)
Comment (BaseInfinity) [reply tests/conftest.py]: Made the change in latest commits to use `tests/cassettes`

**Comment by BaseInfinity** [comment]
Cool I think I ironed out the test issues

**Comment by BaseInfinity** [comment]
I plan to work on this more over the weekend, I'll just use my own git submodule now and we can swap it out with the official git submodule repo towards the end

File: .gitmodules [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[submodule "tests/cassettes"]
+	path = tests/cassettes
+	url = https://github.com/BaseInfinity/cassettes
```
Comment (BaseInfinity) [inline .gitmodules:3]: Ill be sure to change this before merging

**Comment by BaseInfinity** [comment]
I made more changes to get closer to the goal (I need to test my changes to the CI workflow file but im having trouble getting them to stick)

For example I made some changes to the `ci.yml` workflow file and noticed none of my new steps were appearing:
https://github.com/Significant-Gravitas/Auto-GPT/actions/runs/5031267556/jobs/9024235171?pr=4294

Though when I look at the Workflow file it has my changes:
https://github.com/Significant-Gravitas/Auto-GPT/actions/runs/5031267556/workflow?pr=4294#L73-L76

How can I ensure that my CI workflow is being used? 

It looks as though my changes to `docker-ci` workflow is being run so a little confused 🤷 

I'm wondering if this is because I'm technically working through a pull on my repo, and not on the pull from within the repo?

**Comment by merwanehamadi** [comment]
@BaseInfinity the best way to test the ci.yml is to merge it into your master: 
baseInfinity:master
Because this ci.myl can only be used once it's merged, it's a security feature

**Comment by BaseInfinity** [comment]
> @BaseInfinity the best way to test the ci.yml is to merge it into your master: baseInfinity:master Because this ci.myl can only be used once it's merged, it's a security feature

Ahh okay i had a feeling it had to do with something like that, I'll do the thing tomorrow and try again, ty for the tip!

**Comment by k-boikov** [comment]
https://github.com/Significant-Gravitas/Auto-GPT/pull/4355

**Comment by BaseInfinity** [comment]
@merwanehamadi hmm I'm having a hard time getting any of my changes to run for `ci.yml` even in my forked repo:
https://github.com/BaseInfinity/Auto-GPT/actions/runs/5051865442/workflow?pr=3#L77-L78
Screen Shot 2023-05-22 at 5 49 32 PM

I'm gonna see if I can work around this issue in the meantime by duplicating the workflow in my repo and only running the code pertaining to my changes

**Comment by BaseInfinity** [comment]
I was able to create a duplicate CI yaml file in my repo to test my changes on my fork:
https://github.com/BaseInfinity/Auto-GPT/actions/runs/5052043243/jobs/9064500565?pr=3

During the run I printed out the directory of `tests/cassettes` to ensure that the submodule was being loaded
Screen Shot 2023-05-22 at 6 17 30 PM

While the tests didn't completely pass, most did (just a few didn't), so I figured the mechanism to load the cassettes is working
Screen Shot 2023-05-22 at 6 19 40 PM

This is ready for review now

Is there a way better way to test this in this repo to prove it works? Or am I limited to doing all my tests on my fork?

**Comment by BaseInfinity** [comment]
Reminder that I'll want to change this to the actual cassettes repo you guys create:
https://github.com/Significant-Gravitas/Auto-GPT/pull/4294/files#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584R3

**Comment by BaseInfinity** [comment]
Looking at Files Changed might be easiest way to consume this since the pull was pretty simple, I just had to take a lot of mini steps in order to test CI

**Comment by ntindle** [comment]
@merwanehamadi how does this work with your changes?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -70,6 +70,7 @@ jobs:
           fetch-depth: 0
           ref: ${{ github.event.pull_request.head.ref }}
           repository: ${{ github.event.pull_request.head.repo.full_name }}
+          submodules: recursive
```
Comment (Pwuts) [inline .github/workflows/ci.yml:73]:
```suggestion
          submodules: true
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -93,43 +94,21 @@ jobs:
...
-        with:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          message: |
-            Please click [HERE](${{ steps.create_pr.outputs.pull-request-url }}) and merge this PR to update the cassettes.
-
-      - name: Fail if new PR created
-        if:  ${{ env.TIMESTAMP_COMMIT != null }}
-        run: exit 1
+          git add .
+          git commit -am "Updated cassettes submodule"
```
Comment (Pwuts) [inline .github/workflows/ci.yml:105]:
```suggestion
          git commit -m "Auto-update cassettes after merging PR #${{ github.event.pull_request.number }}"
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -93,43 +94,21 @@ jobs:
...
+          git push origin main
+          cd ../..
+          git config user.email "github-actions@github.com"
+          git config user.name "GitHub Actions"
+          git add tests/cassettes
+          git commit -am "Updated main repo's submodules"
```
Comment (Pwuts) [inline .github/workflows/ci.yml:111]:
```suggestion
          git commit -m "Auto-update cassettes submodule after merging PR #${{ github.event.pull_request.number }}"
```

File: .github/workflows/docker-ci.yml [hunk 7]
```diff
@@ -75,8 +75,11 @@ jobs:
     runs-on: ubuntu-latest
     needs: build
     steps:
-    - name: Checkout repository
-      uses: actions/checkout@v3
+      - name: Check out repository
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: 'recursive'
```
Comment (Pwuts) [inline .github/workflows/docker-ci.yml:82]:
```suggestion
          submodules: true
```

**Review by Pwuts — COMMENTED** [review]
Looking very good! I have created https://github.com/Significant-Gravitas/Auto-GPT-test-cassettes and added @BaseInfinity as a maintainer. You can push the cassettes there and set it as the `tests/cassettes` submodule.

To what extent have you tested the CI workflow? The git authentication requires some extra attention because git does not use the `GITHUB_TOKEN` by default.

**Comment by BaseInfinity** [comment]
I bequeath this pull request to @merwanehamadi

**Comment by BaseInfinity** [comment]
Oh ya one last thing to mention that I didn't solve was knowing where to put this command:
`ln -s hooks/post-checkout .git/hooks/post-checkout`

I'm not used to python env but I figure there should be a way to include it as part of the `pip install` process somehow?

**Pull request closed without merging**